### DISCUSSION TITLE
Is it safe to record the offset of a value instead of consuming it right away?

### DISCUSSION THREAD

**Discussion by Gun9niR** [simdjson/simdjson#2205] [Q&A] [answered]
I want to reuse some values in the document, but want to avoid copying those values if at all possible. Is it safe to just compute the offset of the value with `field.value().data() - paddedJson`, and access them later in the json string with this offset?

**Comment by lemire** [comment]
I'm not sure I understand this code (`field.value().data() - paddedJson`). Do you have a complete example of what you want to do?

**Comment by Gun9niR** [comment]
Here's a minimal example of what I'm trying to do

```c++
#include "simdjson.h"
#include <iostream>
#include <string>
#include <vector>

using namespace simdjson;

struct JsonFieldRef {
  size_t offset;
  size_t len;
};

int main() {
  std::string json = "{\"a\":1, \"b\": 2}";
  padded_string json_padded = padded_string(json);
  std::vector<JsonFieldRef> fields;

  ondemand::parser parser;
  auto doc = parser.iterate(json_padded);
  auto object = doc.get_object();
  for (auto field : object) {
    auto value = field.value().raw_json().value();
    auto offset = value.data() - json_padded.data();
    auto len = value.size();
    fields.emplace_back(JsonFieldRef{static_cast<size_t>(offset), len});
  }

  for (auto field_ref : fields) {
    std::cout << std::string_view(json.data() + field_ref.offset, field_ref.len)
              << std::endl;
  }

  return 0;
}
```

Given that I have access to the original Json string all the time, I want to extract the values so I can reuse them later. The catch is I want to do it just by storing their offset and length instead of copying them to a new `std::string`. I'm just wondering if simdjson does anything that would make offset computation `auto offset = value.data() - json_padded.data()` yield the wrong result. My impression is that it doesn't because of the fact that `parser::iterate` takes a `const char*`, and padding the string does not modify the original string.

**Reply by lemire** [reply, marked as answer]
It is safe, but why not just capture string_view instances.

Can you review this update to our documentation:

https://github.com/simdjson/simdjson/pull/2206

**Reply by Gun9niR** [reply]
Right, in this example using string_view is fine, in my real code though when I use the value later it's on a JSON string with a different starting address but same content.
